### ISSUE TITLE
Some Netty features are not enabled by default

### ISSUE THREAD

**Issue by franz1981** [quarkusio/quarkus#39907]
### Description

Currently there are (after the introduction of Java modules), few Netty features, which are not enabled because the "right" JVM args are not present by default on quarkus applications:
1. allocating direct buffer(s) without `Cleaner`s: this is used on off-heap pooled arena(s) backed NIO `ByteBuffer`s.  It can be enabled adding `-Dio.netty.tryReflectionSetAccessible=true --add-opens=java.base/java.nio=ALL-UNNAMED` to the JVM args. 
2. allocating "big enough" (>  1024 by default) `byte[]` without zeroing them, using the "hidden" feature explained at https://shipilev.net/jvm/anatomy-quarks/7-initialization-costs/#_experiment. It can be enabled via ` -Dio.netty.tryReflectionSetAccessible=true --add-opens=java.base/jdk.internal.misc=ALL-UNNAMED`

I'm **not** proposing to add both, but I think is important to be aware that these features are not available in the current default configuration.

The first problem (related the cleaner option) is pretty important, because:
- without configurations as it is now: it implies that the practical (direct) memory limits are the ones specified by Java by default, which means == the heap size (it means that a container should have **2*max Java heap size** available TOTAL memory: max java heap capacity + direct memory limits). Additionally, the JVM is responsible to expose via its metrics (yeeeee!) these direct allocations, and will take care of issuing `OutOfDirectMemory` (which is not "yay", the opposite: see https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/java.base/share/classes/java/nio/Bits.java#L145-L178 how ugly can go!)
- with the proposed configuration: the direct memory limits are still (if nothing is configured, see https://github.com/netty/netty/blob/b0e43847a8c78325ea7dd8bb02934ed6c690cb9c/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L151-L158) the same as the max Java heap **BUT** the JDK NIO direct `ByteBuffer`s allocations are tracked separately by OpenJDK, while the Netty direct allocations would use `Unsafe::allocateMemory`, instead, and its limit are enforced by Netty, instead. This means that the TOTAL memory required (in theory) to a container is: **3*max java heap**(!!!! ie heap + direct NIO memory limit + Netty direct memory limit). Additionally, there won't be any JVM metrics (maybe just native memory tracking AFAIK), which can report the total number of allocated direct memory (including the one from Netty, via Unsafe). Thanks to @alesj we can still use the Netty memory metrics, anyway, for this purpose.

So, why this second option could be appealing?
1. the allocated NIO `ByteBuffer`s won't contain any `Cleaner` saving some memory
2. the default heuristic for the platform, while going OOM, is not great (see https://github.com/openjdk/jdk/blob/jdk-21%2B35/src/java.base/share/classes/java/nio/Bits.java#L145-L178) and can cause an application to become unresponsive 

### Implementation ideas

Configured the proposed option by default

**Comment by franz1981** [comment]
@zakkak @cescoffier

**Comment by cescoffier** [comment]
Hum, we cannot really recommend using `--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED`.

**Comment by zakkak** [comment]
I agree we should only resolve to opening core modules only if strictly necessary.

**Comment by franz1981** [comment]
To me is fine to:
- `-Dio.netty.tryReflectionSetAccessible=true --add-opens=java.base/java.nio=ALL-UNNAMED` and see if it bring evident benefits (on native image)
- just don't do anything and just remember about this note: in the past Wildfly people didn't have such note and even sizing memory on Netty (which is part of this) has been my "most answered question" of the last 4 years, so, still worthy to be aware of it

**Comment by franz1981** [comment]
The issue related zeroing of heap buffers (i.e. byte[]) can indeed be checked (in native image as well), by enabling debug/trace logging and search for these log messages: https://github.com/netty/netty/blob/afb475dabd5ca9eaa94ccf07f0ecd8431db008ba/common/src/main/java/io/netty/util/internal/PlatformDependent0.java#L507-L517

@zakkak

**Comment by cescoffier** [comment]
We would need alternatives to enable these options without requiring an add-open.

**Comment by franz1981** [comment]
We don't need to change anything for my part (I hope):
- SSL requires changing or exposing a new option in the engine (a vertx config object)
- unified allocators requires a sys prop

**Comment by franz1981** [comment]
The alternative for the one impacting SSL is already in; by having SSL heap buffer pooling we don't need to care about allocate uninitialized arrays, cause we won't allocate them anymore 😁 
For the cleaner one, I have to think 🤔

**Comment by franz1981** [comment]
Fyi @cescoffier @zakkak https://blog.csdn.net/jdcdev_/article/details/132843927

This is why we should enable reflective access to create direct buffers without cleaners.
Especially with containers.

Use the Google translator to read it 🙏

**Comment by franz1981** [comment]
Just to add more info, a run of latest techempower plaintext with/without such options shown this improvement

![Image](https://github.com/user-attachments/assets/70c2c50c-2a41-4203-8b73-3aa5bd993a4c)

Both tests used latest JDK 24
The 75473 run just used Quarkus configured OOTB, whilst 75653 got:
```
-Dio.netty.tryReflectionSetAccessible=true --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/jdk.internal.misc=ALL-UNNAMED --sun-misc-unsafe-memory-access=allow
```
and probably some of these are just redundant. I've used the Netty 4.1 guide at https://netty.io/wiki/java-24-and-sun.misc.unsafe.html

And It doesn't account for the improvement we could have in memory usage too (since using Unsafe's variant doesn't allocate any additional Cleaner object).

@cescoffier @geoand

**Comment by zakkak** [comment]
Fixed in https://github.com/quarkusio/quarkus/pull/51807

**Comment by franz1981** [comment]
And just FYI, together with the improvement at https://github.com/quarkusio/quarkus/issues/39907#issuecomment-3003966539 we have (to be verified, but there are no other changes which seems related to this yet!) other improvements on native image RSS e.g.

<img width="1511" height="1126" alt="Image" src="https://github.com/user-attachments/assets/e826946c-475d-4f75-9f70-49030ad14f3e" />

@Karm @Sanne ❤️

**Comment by franz1981** [comment]
To be fair https://github.com/quarkusio/quarkus/pull/51677 as @zakkak found can equally be the reason (and TBH by some margin, more likely too!) of the native image RSS improvements

**Comment by zakkak** [comment]
I performed some measurements on my machine using Quarkus 3.30 that doesn't include https://github.com/quarkusio/quarkus/issues/39907

I used https://github.com/quarkus-qe/quarkus-startstop and there indeed seems to be a reduction in "warm" (after the first request) RSS when using Mandrel 25 (most probably due to https://github.com/oracle/graal/pull/9821).

* JAKARTA_REST_MINIMAL goes from ~46kB down to ~42kB
* FULL_MICROPROFILE goes from ~78kB down to ~68kB

Once Quarkus 3.31 is out I will do another comparison to see the impact of https://github.com/quarkusio/quarkus/issues/39907 as well
